lib/rex/post/meterpreter/packet.rb

```diff
@@ -671,20 +744,99 @@ def initialize(type = nil, method = nil)
     end
   end
 
+  def add_raw(bytes)
+    self.raw << bytes
+  end
+
+  def raw_bytes_required
+    # if we have the xor bytes and length ...
+    if self.raw.length >= PACKET_HEADER_SIZE
+      # return a value based on the length of the data indicated by
+      # the header
+      xor_key = self.raw.unpack('A4')[0]
+      decoded_bytes = xor_bytes(xor_key, raw[0, PACKET_HEADER_SIZE])
+      _, _, _, length, _ = decoded_bytes.unpack('A4A16NNN')
+      length + PACKET_HEADER_SIZE - HEADER_SIZE - self.raw.length
+    else
+      # Otherwise ask for the remaining bytes for the metadata to get the packet length
+      # So we can do the rest of the calculation next time
+      PACKET_HEADER_SIZE - self.raw.length
+    end
+  end
+
+  def aes_encrypt(key, data)
+    # Create the required cipher instance
+    aes = OpenSSL::Cipher.new('AES-256-CBC')
```

Comment:
For later efforts atop this - probably want to make ciphers selectable or even stackable.